### ISSUE TITLE
Poor Memory management in Tensor.diagflat

### ISSUE THREAD

**Issue by et-hicks** [oortega20/tensorslow#1]
https://github.com/oortega20/tensorslow/blob/b096f537aed94a762e044b7b2bf53f16e19cfb8c/src/tensorslow/linalg/tensor.py#L363

Instead of initilizing a new array every time, make a keyword argument in the function that says 

```
def diagflat(cls, data: list, zeros=None):
# ... logic
```

that way you dont have to make zeros a bunch of times each time.

When youre using your derivative methods, you can replace the diagonal arrays across the loops! this way you go from a O(n^2) method each time to a O(n) on average, with a cold start or a worse case of O(n^2)

**Comment by et-hicks** [comment]
change your softmax derivative function from

```python
    def derivative(self, dout: Tensor) -> Tensor:
        if dout and not dout.order == 2:
            raise ValueError('Softmax derivative only for tensors of order 2')

        num_samples, num_classes = self.x.shape
        grad = Tensor([], (num_samples, num_samples), init='zeros')
        for n in range(num_samples):
            sm = Tensor.diagflat(self.x.tensor[n])
            outer_prod = Tensor(self.x.tensor[n], (num_classes, 1))
            grad = grad + (sm - (outer_prod @ outer_prod.T))
        return grad if not dout else dout.T @ grad
```

to

```python
    def derivative(self, dout: Tensor) -> Tensor:
        if dout and not dout.order == 2:
            raise ValueError('Softmax derivative only for tensors of order 2')

        num_samples, num_classes = self.x.shape
        grad = Tensor([], (num_samples, num_samples), init='zeros')
        sm = Tensor.zeroes(n, n) # idk if this exists
        for n in range(num_samples):
            sm = Tensor.diagflat(self.x.tensor[n], zeroes=sm) # once you alter this method
            outer_prod = Tensor(self.x.tensor[n], (num_classes, 1))
            grad = grad + (sm - (outer_prod @ outer_prod.T))
        return grad if not dout else dout.T @ grad
```

**Comment by et-hicks** [comment]
closed with ea838c65c389e1604110cb720222a5c14ba97495
